Add support for ETag and Last-Modified headers for summary and summary.sig

pwithnall opened the pull request:
See the commit messages. This adds support (in the soup and curl backends) for the `Last-Modified`/`ETag`/`If-None-Match`/`If-Modified-Since` HTTP headers and hooks them up for requests for `summary` and `summary.sig` so they don’t have to be re-downloaded if unchanged from last time. In particular, this will shave 590B off each response for `summary.sig` in the cached (common) case.

This PR is related to #2167 in that it’s a network efficiency improvement, but can be reviewed/merged independently.

pwithnall commented:
One possible improvement to think about after this (if it gets merged) is that the separate HTTP requests to get `summary.sig` and then (if it’s changed) `summary` could be combined: previously `summary.sig` was essentially acting as the `ETag` for `summary`, but once HTTP ETags are supported properly, then perhaps that could be rethought. If a client has an ETag for its cached `summary`, it could request `summary` from the server immediately (without requesting `summary.sig` first) and rely on the ETag it sends to make sure it only downloads the full `summary` file when it’s actually changed.

That would mean a shift to using inline signatures to protect `summary`, though (otherwise you still need to download `summary.sig` separately). I can’t remember all of the tradeoffs to know whether that’s helpful, unhelpful, or indifferent.

jlebon reviewed (commented):
Overall idea makes sense to me. Though worth noting there was a suggestion from @alexlarsson recently to have deltas for summary files as well which I think can affect the discussion here. E.g. if we were to support summary deltas, then presumably they'd be in some form which makes querying for "new summary wrt what I have locally" cheap, which intersects with what's going on here.

Though support for cache semantics makes sense on its own too and is compatible with existing OSTree webservers today.

pwithnall commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +590,179 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+ *    Wed, 21 Oct 2015 07:28:00 GMT
+ */
+static GDateTime *
+parse_rfc2616_date_time (const char *buf,
```
There’s `soup_date_new_from_string()`, but it’s in libsoup and this code is in the libcurl backend, so I assume libsoup is not an option there. Have I misunderstood?

cgwalters replied:
Another big thing to keep in mind is that when updating host systems at least ostree usually runs as unrestricted root (I'd eventually like to fix this) and things like the string parsing we're doing here are an attack surface and need careful auditing.

To be clear did you write this code from scratch or is it inspired by anything else?

Let's at least add some unit tests for this? 

We've already copied in some libsoup code for URIs, maybe we could copy in the libsoup code?

pwithnall replied:
> To be clear did you write this code from scratch or is it inspired by anything else?

It’s written from scratch.

> Let's at least add some unit tests for this?

I’ll have a go.

> We've already copied in some libsoup code for URIs, maybe we could copy in the libsoup code?

Maybe, but it was [rewritten recently](https://gitlab.gnome.org/GNOME/libsoup/-/commit/4596b4ec4c14a0e473e3fa0c3f41a0107d846cc8), so is in basically the same position as this code (except more general and has support for multiple RFC date formats tied together in the same code). The older version of the code was tied to `SoupDate` and also supported multiple date formats in one code path.

I think having a minimal in-house implementation makes sense, given that we only need to support the `Last-Modified` header, which only uses one specific HTTP date format (RFC 2616). If we were supporting more headers with other formats then maybe things would be different.

cgwalters replied:
Thanks so much for the added tests!

pwithnall commented on src/libostree/ostree-repo-pull.c:
```diff
@@ -2700,6 +2701,66 @@ _ostree_repo_verify_summary (OstreeRepo   *self,
+
+  if (out_etag != NULL)
+    {
+      g_autoptr(GBytes) etag_bytes = NULL;
+
+      etag_bytes = glnx_fgetxattr_bytes (fd, "user.etag", NULL);
```
So declarations not necessarily at the start of code blocks? I thought there were problems with doing that with autocleanup, where the autocleanup would happen but the initialisation might not (if not reached by the time the block is exited)?

pwithnall replied:
OK. Tweaked a couple of declarations to move them to where they’re used, but most declarations in the PR are assigned by pointer so can’t be changed.

pwithnall commented on src/libostree/ostree-repo-pull.c:
```diff
@@ -2700,6 +2701,66 @@ _ostree_repo_verify_summary (OstreeRepo   *self,
+            *out_etag = NULL;
+        }
+      else
+        *out_etag = NULL;
```
Could do, although it would likely get missed if this function was ever expanded to do error handling. I can change it if you want.

pwithnall commented on src/libostree/ostree-repo-pull-private.h:
```diff
@@ -72,7 +72,11 @@ typedef struct {
   gboolean          has_tombstone_commits;
 
   GBytes           *summary_data;
+  char             *summary_etag;
+  guint64           summary_last_modified;  /* seconds since the epoch */
```
I think `ETag` support is actually more important than `Last-Modified` support because it doesn’t have race conditions around updates/checks in the same second, and doesn’t have issues with clock synchronisation. Both require some level of support on the server (typically, aiui, servers are configured to serve everything as if it’s not cacheable, as that’s the safest option). I would expect that all web servers would support both headers, though. They’ve been around a long time.

If xattr support isn’t available then ETags won’t work, but should degrade nicely, so that shouldn’t cause a problem.

jlebon commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -335,7 +341,16 @@ check_multi_info (OstreeFetcher *fetcher)
           curl_easy_getinfo (easy, CURLINFO_RESPONSE_CODE, &response);
-          if (!is_file && !(response >= 200 && response < 300))
+
+          if (!is_file && response == 304)
+            {
+              /* Version on the server is unchanged from the version we have
+               * cached locally; report this as an out-argument, a zero-length
+               * response buffer, and no error. */
+              req->out_not_modified = TRUE;
+            }
```
Optional: should we check here that at least one of `if_none_match` or `if_modified_since` was provided? That way, if somehow a HTTP server returns 304 even though we didn't provide either of these, we still error out as before.

pwithnall replied:
Good suggestion, will tweak things to do that (and in the libsoup implementation too).

jlebon reviewed (commented):
Would be nice to have a basic test for this where we check that we avoid pulling the summary. Something like just doing another `ostree pull` and sanity-checking that the inode for the summary didn't change? (Since we always create a new file when caching.) Hmm, though on the `ostree-trivial-httpd` side, briefly looking at `SoupServer`, it seems like we'd have to implement that logic ourselves like [this test here](https://gitlab.gnome.org/GNOME/libsoup/-/blob/4596b4ec4c14a0e473e3fa0c3f41a0107d846cc8/tests/cache-test.c#L55-68).

pwithnall commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +590,179 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+  const char *end_ptr = NULL;
+  gint saved_errno = 0;
+
+  g_return_val_if_fail (n_digits == 2 || n_digits == 4, FALSE);
```
Whoops, good catch. It was used before I refactored the function a bit, and then I forgot to plumb it back in. Will update.

pwithnall commented:
> Would be nice to have a basic test for this where we check that we avoid pulling the summary

Working on that now; will hopefully have something ready tomorrow.

pwithnall commented:
Updated. I’ve fixed a couple of bugs and added a unit test (although see its commit message: it doesn’t actually fail with the ETag/Last-Modified changes reverted). All my changes are as fixups for ease of review, and can be squashed later.

cgwalters commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +591,178 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+      strncasecmp (buffer, "ETag: ", strlen ("ETag: ")) == 0)
+    {
+      g_clear_pointer (&req->out_etag, g_free);
+      req->out_etag = g_strstrip (g_strdup (buffer + strlen ("ETag: ")));
```
`g_strstrip()` is dangerous here because if the buffer has *leading* whitespace we lose track of the real malloc pointer.  I think you just want `g_strchomp()`?

pwithnall replied:
Are you sure? `g_strstrip(x)` is just `g_strchug(g_strchomp(x))`, and `g_strchug()` doesn’t increment the input pointer, [it uses `memmove()`](https://gitlab.gnome.org/GNOME/glib/-/blob/master/glib/gstrfuncs.c#L2284) to shift the string contents down over any leading whitespace. The pointer is unchanged.

cgwalters commented on src/libostree/ostree-repo-pull.c:
```diff
@@ -2772,11 +2834,38 @@ _ostree_repo_load_cache_summary_if_same_sig (OstreeRepo        *self,
+  glnx_autofd int fd = -1;
+  struct timeval time_vals[] =
+    {
+      { .tv_sec = last_modified, .tv_usec = 0 },  /* access */
```
Why reset the access time versus `UTIME_OMIT`?

pwithnall replied:
Because I didn’t know about `UTIME_OMIT`. Fixed, thanks

cgwalters commented on src/libostree/ostree-fetcher-soup.c:
```diff
@@ -431,6 +438,24 @@ create_pending_soup_request (OstreeFetcherPendingURI  *pending,
+      g_autoptr(GDateTime) date_time = NULL;
+
+      date_time = g_date_time_new_from_unix_utc (pending->if_modified_since);
+      mod_date = g_date_time_format (date_time, "%a, %d %b %Y %H:%M:%S %Z");
```
This kind of stuff I think would look more readable in declare-and-initialize style.  (Not a blocker though)

pwithnall replied:
Fixed

cgwalters commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +591,178 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+static size_t
+response_header_cb (const char *buffer, size_t size, size_t n_items, void *user_data)
+{
+  const size_t real_size = size * n_items;
```
Does libcurl protect us against overflow?  Hmm, [the docs](https://curl.haxx.se/libcurl/c/CURLOPT_HEADERFUNCTION.html) say that "size is always one" so I guess so.

pwithnall replied:
Yeah, it’s a bit of a weird API. I’ll add an assertion to be sure

cgwalters commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +591,178 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+      req->out_etag = g_strstrip (g_strdup (buffer + strlen ("ETag: ")));
+    }
+  else if (real_size > strlen ("Last-Modified: ") &&
+           strncasecmp (buffer, "Last-Modified: ", strlen ("Last-Modified: ")) == 0)
```
I think this would read more nicely with a `static char last_modified[] = "Last-Modified: "` somewhere above, would also protect against tyops.  Same for etag.

pwithnall replied:
True, fixed

cgwalters commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +591,178 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+      else
+        dt = parse_rfc2616_date_time (lm_buf, lm_len - 2);
+
+      req->out_last_modified = (dt != NULL) ? g_date_time_to_unix (dt) : 0;
```
Do we need an `if (req->out_last_modified) g_date_time_unref (req->out_last_modified);` here?

pwithnall replied:
No, `req->out_last_modified` is a `guint64`

cgwalters commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +591,178 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+      const char *lm_buf = buffer + strlen ("Last-Modified: ");
+      size_t lm_len = real_size - strlen ("Last-Modified: ");
+
+      /* Check termination */
```
Why do we do this in the last-modified case but not etag?

pwithnall replied:
Simplicity of the code; stripping the termination is handled in the ETag case by `g_strstrip()`, and it seemed overkill to check that the last two bytes actually were `\r\n`, given that the set of allowed characters in an ETag is basically everything, and that all we do with it is to store it in an xattr and then regurgitate it. My thinking was that as long as we’re not actually parsing the ETag (and correctly handle it as an arbitrary byte array) it doesn’t need validating, just storing and regurgitating.

I can tighten up the parsing on the ETag if you want, though; I think the HTTP spec is not 100% liberal with it.

jlebon replied:
Looking at https://stackoverflow.com/a/50180173, it seems like it's possible to have arbitrary amounts of whitespace before the value as well. And that'll throw off the `len != 29` check in the parser. So maybe simpler to just `g_strstrip (g_strdup (...))` here too.

pwithnall replied:
Done

jlebon commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -721,8 +785,32 @@ initiate_next_curl_request (FetcherRequest *req,
+
+  if (req->if_modified_since > 0)
+    {
+      g_autofree char *mod_date = NULL;
+      g_autoptr(GDateTime) date_time = NULL;
+
+      date_time = g_date_time_new_from_unix_utc (req->if_modified_since);
+      mod_date = g_date_time_format (date_time, "If-Modified-Since: %a, %d %b %Y %H:%M:%S %Z");
```
Could use new style here. ;)

jlebon commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -335,7 +342,17 @@ check_multi_info (OstreeFetcher *fetcher)
+               * cached locally; report this as an out-argument, a zero-length
+               * response buffer, and no error. */
+              req->out_not_modified = TRUE;
+            }
+
+          if (!is_file && !(response >= 200 && response < 300) && response != 304)
```
Minor: I think it'd be nice to match the libsoup version of this and use `else if` here. I like that it makes it really obvious that if the first branch is taken, it's not an error condition.

pwithnall replied:
I tried that and it breaks fetches if done in libcurl. Looking more closely, the structure of the code doesn’t match: in libsoup, the `if` block is entirely for error handling, but in libcurl, the first `if` block is for error handling and the subsequent `else if` and `else` blocks are all for success handling. We want to take one of those `else if`/`else` blocks even if `req->out_not_modified` has been set.

So I haven’t made this change.

jlebon commented on tests/test-rfc2616-dates.c:
```diff
@@ -0,0 +1,123 @@
+      const char *rfc2616;
+      const char *expected_iso8601;  /* (nullable) if parsing is expected to fail */
+    }
+  tests[] =
```
Nice!

jlebon reviewed (approved):
This looks great overall! Really just minor things.

Could you squash the commits? We don't use the Homu bot here anymore which means that there is no autosquashing.

pwithnall commented on src/libostree/ostree-date-utils.c:
```diff
@@ -0,0 +1,166 @@
+    return NULL;
+  if (*(month_name + 3) != ' ')
+    return NULL;
+  if (!parse_uint (year, 4, 0, 9999, &year_int))
```
I like to aim higher than [Busted](https://www.youtube.com/watch?v=Tu7HoGZaspo) in all that I do.

pwithnall commented on src/libostree/ostree-fetcher-curl.c:
```diff
@@ -575,6 +592,50 @@ write_cb (void *ptr, size_t size, size_t nmemb, void *data)
+
+  /* libcurl says that @size is always 1, but let’s check
+   * See https://curl.haxx.se/libcurl/c/CURLOPT_HEADERFUNCTION.html */
+  g_assert (size == 1);
```
No, `g_assert()` is appropriate here. The `g_assert_*()` functions are intended for use in unit tests, and [have all sorts of test-specific behaviour](https://gitlab.gnome.org/GNOME/glib/-/blob/master/glib/gtestutils.c#L2885). `g_assert()` does not, and is correctly compiled out if OSTree were to be built with `G_DISABLE_ASSERT`. (I don’t advise doing that, though, it makes debugging real-world bugs a pain.)

jlebon replied:
Hmm, we use `g_assert_*()` macros a lot in this codebase (and [rpm-ostree](https://github.com/coreos/rpm-ostree) too).

> The `g_assert_*()` functions are intended for use in unit tests, and have all sorts of test-specific behaviour. `g_assert()` does not, and is correctly compiled out if OSTree were to be built with `G_DISABLE_ASSERT`.

Ahh OK, so the primary difference you're pointing out here is that `g_assert()` is elided with `G_DISABLE_ASSERT` while the other ones are not. Which is true, though OTOH... I don't think anyone is going to come out and tell us that these pesky assertions are blocking them from the undefined behaviour they're looking for :).

Anyway, fine as is too of course. Just wanted to mention they're used extensively elsewhere too.

pwithnall replied:
> Just wanted to mention they're used extensively elsewhere too.

Technically OSTree is in the wrong there, but in practical terms if this hasn’t caused problems already then it’s unlikely to :)

pwithnall commented:
All review comments addressed (or commented on) and all the fixups squashed.

cgwalters reviewed (approved):
/lgtm

cgwalters commented:
Looks like Travis is stuck, hmm.  Not sure how to retrigger that one without another force push.  Let's give it a bit and see if it comes back, otherwise we can override.
